src/lib/comms/sol-http-client-impl-curl.c

```diff
@@ -746,20 +639,31 @@ static bool
 set_cookies_from_params(CURL *curl, struct sol_arena *arena,
     const struct sol_http_param *params)
 {
-    char *cookies = build_cookies(curl, params);
-
-    if (cookies && !*cookies) {
-        free(cookies);
-        return true;
-    }
-    return set_string_option(curl, CURLOPT_COOKIE, arena, cookies);
+    struct sol_buffer buf;
+    char *cookies;
+    bool r = true;
+    int err;
+
+    sol_buffer_init(&buf);
+    err = sol_http_encode_params(&buf, SOL_HTTP_PARAM_COOKIE, params);
+    SOL_INT_CHECK(err, < 0, false);
+
+    cookies = sol_buffer_steal_or_copy(&buf, NULL);
+    if (cookies)
+        r = set_string_option(curl, CURLOPT_COOKIE, arena, cookies);
+    sol_buffer_fini(&buf);
```

Comment:
Just checked and strings are already copied by cURL, we do not have to keep them in our arena, see:

http://curl.haxx.se/libcurl/c/curl_easy_setopt.html

Thus we really should revisit it, stop calling "steal or copy".